webapp/src/components/backstage/playbook_runs/playbook_run/retrospective.tsx

```diff
@@ -1,13 +1,258 @@
+import {AdminNotificationType} from 'src/constants';
+import {Timestamp} from 'src/webapp_globals';
+import {AnchorLinkTitle, Content, ELAPSED_TIME} from 'src/components/backstage/playbook_runs/shared';
+import MetricsData from '../playbook_run_backstage/metrics/metrics_data';
+import Report from '../playbook_run_backstage/retrospective/report';
+import ConfirmModalLight from 'src/components/widgets/confirmation_modal_light';
+import {PrimaryButton} from 'src/components/assets/buttons';
+
+interface Props {
+    playbookRun: PlaybookRun;
+    setPlaybookRun: (value: React.SetStateAction<PlaybookRun | null>) => void
+}
+
+const editDebounceDelayMilliseconds = 2000;
+
+const Retrospective = ({
+    playbookRun,
+    setPlaybookRun,
+}: Props) => {
+    const [playbook, setPlaybook] = useState<PlaybookWithChecklist | null>(null);
+    const allowRetrospectiveAccess = useAllowRetrospectiveAccess();
+    const {formatMessage} = useIntl();
+    const [showConfirmation, setShowConfirmation] = useState(false);
+    const childRef = useRef<any>();
+    const metricsAvailable = useAllowPlaybookAndRunMetrics();
+
+    useEffect(() => {
+        const fetchData = async () => {
+            if (playbookRun?.playbook_id) {
+                const fetchedPlaybook = await clientFetchPlaybook(playbookRun.playbook_id);
+                setPlaybook(fetchedPlaybook ?? null);
+            }
+        };
+
+        fetchData();
+    }, [playbookRun?.playbook_id]);
+
+    if (!allowRetrospectiveAccess) {
+        return (
+            <Container>
+                <AnchorLinkTitle
+                    title={formatMessage({defaultMessage: 'Retrospective'})}
+                    id='retrospective'
+                />
+                <BannerWrapper>
+                    <UpgradeBanner
+                        background={<></>}
+                        titleText={formatMessage({defaultMessage: 'Publish retrospective report and access the timeline'})}
+                        helpText={formatMessage({defaultMessage: 'Celebrate success and learn from mistakes with retrospective reports. Filter timeline events for process review, stakeholder engagement, and auditing purposes.'})}
+                        notificationType={AdminNotificationType.RETROSPECTIVE}
+                        verticalAdjustment={0}
+                        vertical={true}
+                    />
+                </BannerWrapper>
+            </Container>
+        );
+    }
+
+    const confirmedPublish = () => {
+        publishRetrospective(playbookRun.id, playbookRun.retrospective, playbookRun.metrics_data);
+        setPlaybookRun((run) => ({
+            ...run,
+            retrospective_published_at: DateTime.now().valueOf(),
+            retrospective_was_canceled: false,
+        } as PlaybookRun));
+        setShowConfirmation(false);
+    };
+
+    const publishButtonText: React.ReactNode = formatMessage({defaultMessage: 'Publish'});
```

Comment:
isn't the type implicit?

```suggestion
    const publishButtonText = formatMessage({defaultMessage: 'Publish'});
```